src/containers/cyclicbuffer.d

```diff
@@ -0,0 +1,828 @@
+/**
+ * Cyclic Buffer
+ * Copyright: © 2016 Economic Modeling Specialists, Intl.
+ * Authors: Nickolay Bukreyev
+ * License: $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ */
+
+module containers.cyclicbuffer;
+
+private import core.exception : onRangeError;
+private import std.experimental.allocator.mallocator : Mallocator;
+private import std.range.primitives : empty, front, back, popFront, popBack;
+private import containers.internal.node : shouldAddGCRange;
+
+/**
+ * Array that provides constant time (amortized) appending and popping
+ * at either end, as well as random access to the elements.
+ *
+ * Params:
+ *     T = the array element type
+ *     Allocator = the allocator to use. Defaults to `Mallocator`.
+ *     supportGC = true if the container should support holding references to GC-allocated memory.
+ */
+struct CyclicBuffer(T, Allocator = Mallocator, bool supportGC = shouldAddGCRange!T)
+{
+	@disable this(this);
+
+	private import std.conv : emplace;
+	private import std.experimental.allocator.common : stateSize;
+	private import std.traits : isImplicitlyConvertible, hasElaborateDestructor;
+
+	static if (stateSize!Allocator != 0)
+	{
+		/// No default construction if an allocator must be provided.
+		@disable this();
+
+		/**
+		 * Use the given `allocator` for allocations.
+		 */
+		this(Allocator allocator) nothrow pure @safe @nogc
+		in
+		{
+			assert(allocator !is null, "Allocator must not be null");
+		}
+		body
+		{
+			this.allocator = allocator;
+		}
+	}
+
+	~this()
+	{
+		clear();
+		static if (useGC)
+		{
+			import core.memory : GC;
+			GC.removeRange(storage.ptr);
+		}
+		allocator.deallocate(storage);
+	}
+
+	/**
+	 * Removes all contents from the buffer.
+	 */
+	void clear()
+	{
+		if (!empty)
+		{
+			static if (hasElaborateDestructor!T)
+			{
+				if (start <= end)
+					foreach (ref item; storage[start .. end + 1])
+						.destroy(item);
+				else
+				{
+					foreach (ref item; storage[start .. $])
+						.destroy(item);
+					foreach (ref item; storage[0 .. end + 1])
+						.destroy(item);
+				}
+			}
+			start = (end + 1) % capacity;
+			_length = 0;
+		}
+	}
+
+	/**
+	 * Ensures capacity is at least as large as specified.
+	 */
+	size_t reserve(size_t newCapacity)
+	{
+		immutable oldCapacity = capacity;
+		if (newCapacity <= oldCapacity)
+			return oldCapacity;
+		auto old = storage;
+		if (oldCapacity == 0)
+			storage = cast(typeof(storage)) allocator.allocate(newCapacity * T.sizeof);
+		else
+		{
+			auto a = cast(void[]) old;
+			allocator.reallocate(a, newCapacity * T.sizeof);
+			storage = cast(typeof(storage)) a;
+		}
+		static if (useGC)
+		{
+			import core.memory : GC;
+			//Add, then remove. Exactly in that order.
+			GC.addRange(storage.ptr, newCapacity * T.sizeof);
+			GC.removeRange(old.ptr);
+		}
+		if (empty)
+			end = (start - 1 + capacity) % capacity;
```

Comment:
Actually, it doesn't matter - the buffer will be empty anyway.

When `end` is one item prior to `start` (`start == (end + 1) % capacity`, or `end == (start - 1 + capacity) % capacity`, which is the same), that means, the buffer is either empty or completely full. However, when `storage is null` and therefore `capacity == 0`, this invariant is broken (due to division by zero), so there is no way to assign any meaningfull value to `end` upon initialization. Either `start` or `end` must be corrected on the first memory allocation. I decided to correct `end` so that a buffer starts filling with index 0, not 1... Well, it doesn't seem to be a good reason for me any more.

`clear()` may reset either `start` or `end` (or even both), and the only reason to stick to the first option is that it requires one machine instruction less.